src/SDK/Logs/Processor/BatchLogsProcessor.php

```diff
@@ -0,0 +1,273 @@
+                $observer->observe($processed, self::ATTRIBUTES_PROCESSED);
+                $observer->observe($dropped, self::ATTRIBUTES_DROPPED);
+            }, true);
+        $meter
+            ->createObservableUpDownCounter(
+                'otel.logs.log_processor.queue.limit',
+                '{logs}',
+                'The queue size limit',
+            )
+            ->observe(function (ObserverInterface $observer): void {
+                $observer->observe($this->maxQueueSize, self::ATTRIBUTES_PROCESSOR);
+            }, true);
+        $meter
+            ->createObservableUpDownCounter(
+                'otel.logs.log_processor.queue.usage',
+                '{logs}',
+                'The current queue usage',
+            )
+            ->observe(function (ObserverInterface $observer): void {
+                $queued = $this->queue->count() * $this->maxExportBatchSize + count($this->batch);
+                $pending = $this->queueSize - $queued;
+                $free = $this->maxQueueSize - $this->queueSize;
+
+                $observer->observe($queued, self::ATTRIBUTES_QUEUED);
+                $observer->observe($pending, self::ATTRIBUTES_PENDING);
+                $observer->observe($free, self::ATTRIBUTES_FREE);
+            }, true);
+    }
+
+    public function onEmit(ReadWriteLogRecord $record, ?ContextInterface $context = null): void
+    {
+        if ($this->closed) {
+            return;
+        }
+
+        if ($this->queueSize === $this->maxQueueSize) {
+            $this->dropped++;
+
+            return;
+        }
+
+        $this->queueSize++;
+        $this->batch[] = $record;
+        $this->nextScheduledRun ??= $this->clock->now() + $this->scheduledDelayNanos;
+
+        if (count($this->batch) === $this->maxExportBatchSize) {
+            $this->enqueueBatch();
+        }
+        if ($this->autoFlush) {
+            $this->flush();
+        }
+    }
+
+    public function forceFlush(?CancellationInterface $cancellation = null): bool
+    {
+        if ($this->closed) {
+            return false;
+        }
+
+        return $this->flush(__FUNCTION__, $cancellation);
+    }
+
+    public function shutdown(?CancellationInterface $cancellation = null): bool
+    {
+        if ($this->closed) {
+            return false;
+        }
+
+        $this->closed = true;
+
+        return $this->flush(__FUNCTION__, $cancellation);
+    }
+
+    private function flush(?string $flushMethod = null, ?CancellationInterface $cancellation = null): bool
+    {
+        if ($flushMethod !== null) {
+            $flushId = $this->batchId + $this->queue->count() + (int) (bool) $this->batch;
+            $this->flush->enqueue([$flushId, $flushMethod, $cancellation, !$this->running, Context::getCurrent()]);
+        }
+
+        if ($this->running) {
+            return false;
+        }
+
+        $success = true;
+        $exception = null;
+        $this->running = true;
+
+        try {
+            for (;;) {
```

Comment:
nitpick (non-blocking): Is there any other way to flush better than this infinite loop ?

Did you do that in case of messages comes during the flush ?

Reply:
I believe that's the reason - this batch exporter is mostly copied from the traces batch exporter, which I don't completely understand...